### PR TITLE
feat: ai-proxy: add filter erda-auth

### PR DESCRIPTION
**Pull request by dspo**
#### What this PR does / why we need it:
feat: ai-proxy: add filter erda-auth

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](paste your link here)

#### Specified Reviewers:

/assign @your-reviewer

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |  feat: ai-proxy: add filter erda-auth            |
| 🇨🇳 中文    | 增加 erda-auth 插件             |

#### Need cherry-pick to release versions?

Add comment like `/cherry-pick release/1.0` when this PR is merged.

> For details on the cherry pick process, see the [cherry pick requests](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md#how-to-cherry-pick-a-merged-pr) section under [CONTRIBUTING.md](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: api/proto/proto_fetcher.sh [hunk 1]
```diff
@@ -6,7 +6,7 @@ cd "$(dirname "${BASH_SOURCE[0]}")"
 EXTERNAL_REPO_DIR="externalrepo"
 
 # otel remote proto conf
-OTEL_PROTO_REPO="https://github.com/open-telemetry/opentelemetry-proto.git"
+OTEL_PROTO_REPO="https://ghproxy.com/https://github.com/open-telemetry/opentelemetry-proto.git"
```
Comment (sfwn) [inline api/proto/proto_fetcher.sh:9]: no need proxy here

File: cmd/ai-proxy/bootstrap.yml [hunk 2]
```diff
@@ -10,6 +10,7 @@ erda.app.ai-proxy:
   routesRef: conf/routes.yml
   providersRef: conf/providers.yml
   logLevel: ${ env.LOG_LEVEL:debug }
+  erdaOpenapi: ${ env.ERDA_OPENAPI:https://openapi.erda.cloud }
```
Comment (sfwn) [inline cmd/ai-proxy/bootstrap.yml:13]: why is the format not like this: `erdaOpenapi: ${ERDA_OPENAPI:https://openapi.erda.cloud}`

File: internal/apps/ai-proxy/.http [hunk 3]
```diff
@@ -1,68 +1,85 @@
...
-  "contextLength": 5,
-  "source": "erda",
-  "model": "gpt-35-turbo-0301",
-  "temperature": 0.7
+  "name": "erda.cloud",
+  "platform": "erda",
+  "description": "credential for erda.cloud AI Assistant",
+  "enabled": true,
+  "providerName": "azure",
+  "providerInstanceId": "terminus2"
```
Comment (sfwn) [inline internal/apps/ai-proxy/.http:20]: desensitize

File: internal/apps/ai-proxy/common/on.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+type On struct {
+	Key      string `json:"key" yaml:"key"`
+	Operator string `json:"operator" yaml:"operator"`
+	Value    string `json:"value" yaml:"value"`
+}
+
+func (on *On) On(header http.Header) (bool, error) {
+	switch on.Operator {
+	case "exist":
```
Comment (sfwn) [inline internal/apps/ai-proxy/common/on.go:31]: EqualFold

File: internal/apps/ai-proxy/common/on_test.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+	"gopkg.in/yaml.v3"
+
+	"github.com/erda-project/erda/internal/apps/ai-proxy/common"
+)
+
+func TestOn_On(t *testing.T) {
+	var s = `
+key: X-Ai-Proxy-Source
+operator: =
```
Comment (sfwn) [inline internal/apps/ai-proxy/common/on_test.go:29]: also test `exist` case.

File: internal/apps/ai-proxy/filters/azure-director/filter.go [hunk 6]
```diff
@@ -268,7 +267,6 @@ func (f *AzureDirector) AllDirectors() map[string]func(ctx context.Context) erro
 	valueOf := reflect.ValueOf(f)
 	doNothing, _ := typeOf.MethodByName("DoNothing")
 	for i := 0; i < typeOf.NumMethod(); i++ {
-		fmt.Printf("type: %v, typeOfProcessor: %v\n", typeOf.Method(i).Type, doNothing.Type)
```
Comment (sfwn) [inline internal/apps/ai-proxy/filters/azure-director/filter.go:271]: can use debug level

File: internal/apps/ai-proxy/handlers/handler_sessions.go [hunk 7]
```diff
@@ -182,23 +165,24 @@ func (s *SessionsHandler) DeleteSession(ctx context.Context, req *pb.LocateSessi
...
+
+	// try to get request source
+	source, ok := getSource(ctx, req)
+	if !ok {
+		return nil, InvalidSessionSource
 	}
-	if source != "" {
-		where["source"] = source
+
+	var where = map[string]any{
```
Comment (sfwn) [inline internal/apps/ai-proxy/handlers/handler_sessions.go:181]: define a struct

File: internal/apps/ai-proxy/handlers/handler_sessions.go [hunk 8]
```diff
@@ -222,3 +203,34 @@ func (s *SessionsHandler) GetSession(ctx context.Context, req *pb.LocateSessionC
 	}
 	return s.Dao.GetSession(req.GetId())
 }
+
+func getUserId(ctx context.Context, req interface{ GetUserId() string }) (string, bool) {
```
Comment (sfwn) [inline internal/apps/ai-proxy/handlers/handler_sessions.go:207]: ID, the same as others

File: internal/apps/ai-proxy/handlers/handler_sessions.go [hunk 9]
```diff
@@ -222,3 +203,34 @@ func (s *SessionsHandler) GetSession(ctx context.Context, req *pb.LocateSessionC
...
+	if source := req.GetSource(); source != "" {
+		return source, true
+	}
+	if source := apis.GetHeader(ctx, vars.XAIProxySource); source != "" {
+		return source, true
+	}
+	return "", false
+}
+
+func getSessionId(ctx context.Context, req interface{ GetSessionId() string }) (string, bool) {
```
Comment (sfwn) [inline internal/apps/ai-proxy/handlers/handler_sessions.go:230]: ID

File: internal/apps/ai-proxy/provider.go [hunk 10]
```diff
@@ -76,23 +77,33 @@ var (
 			return new(provider)
 		},
 	}
+	rootKeyAuth = transport.WithInterceptors(func(h interceptor.Handler) interceptor.Handler {
+		return func(ctx context.Context, req interface{}) (interface{}, error) {
+			if auth := transport.ContextHeader(ctx).Get("Authorization"); len(auth) == 0 || auth[0] != "Bearer "+os.Getenv("AI_PROXY_ROOT_KEY") {
```
Comment (sfwn) [inline internal/apps/ai-proxy/provider.go:82]: too many magic text

File: internal/apps/ai-proxy/provider.go [hunk 11]
```diff
@@ -175,6 +195,15 @@ func (p *provider) RegisterService(desc *grpc.ServiceDesc, impl interface{}) {
 	}
 }
 
+func (p *provider) initConfig() {
+	for _, s := range []*string{
+		&p.Config.LogLevel,
+		&p.Config.ErdaOpenapi,
+	} {
+		parseEnvExpr(s)
```
Comment (sfwn) [inline internal/apps/ai-proxy/provider.go:203]: Seems unnecessary. Use Infra's mechanism instead.

File: internal/apps/ai-proxy/providers/metrics/collector.go [hunk 12]
```diff
@@ -82,14 +84,37 @@ type lvsDistincter struct {
 }
 
 var mockedAudit = models.AIProxyFilterAudit{
```
Comment (sfwn) [inline internal/apps/ai-proxy/providers/metrics/collector.go:86]: maybe `mockedAudit` should be put at `_test.go` files.

File: internal/apps/ai-proxy/vars/consts.go [hunk 13]
```diff
@@ -15,25 +15,30 @@
...
-	XErdaAIProxySource          = "X-Erda-AI-Proxy-Source"
-	XErdaAIProxyName            = "X-Erda-AI-Proxy-Name"
-	XErdaAIProxyPhone           = "X-Erda-AI-Proxy-Phone"
-	XErdaAIProxyJobNumber       = "X-Erda-AI-Proxy-JobNumber"
-	XErdaAIProxyEmail           = "X-Erda-AI-Proxy-Email"
-	XErdaAIProxyDingTalkStaffID = "X-Erda-AI-Proxy-DingTalkStaffID"
-	XErdaAIProxyPrompt          = "X-Erda-AI-Proxy-Prompt"
-	XAIProxyProvider            = "X-AI-Proxy-Provider"
-	XAIProxyProviderInstance    = "X-AI-Proxy-Provider-Instance"
+	XAIProxySessionId          = "X-AI-Proxy-SessionId"
```
Comment (sfwn) [inline internal/apps/ai-proxy/vars/consts.go:18]: ID

File: internal/apps/ai-proxy/vars/consts_test.go [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+import (
+	"context"
+	"testing"
+
+	"github.com/erda-project/erda/internal/apps/ai-proxy/vars"
+	"github.com/erda-project/erda/pkg/reverseproxy"
+)
+
+func TestContext(t *testing.T) {
+	var ctx = context.Background()
```
Comment (sfwn) [inline internal/apps/ai-proxy/vars/consts_test.go:26]: nothing tested

File: internal/pkg/ai-proxy/route/route.go [hunk 15]
```diff
@@ -90,7 +90,10 @@ func (r *Route) HandlerWith(ctx context.Context, kvs ...any) http.HandlerFunc {
 	ctx = context.WithValue(ctx, reverseproxy.LoggerCtxKey{}, l)
 
 	// set key-values to context
-	for i := 0; i+1 < len(kvs); i++ {
+	for i := 0; i+1 < len(kvs); i += 2 {
```
Comment (sfwn) [inline internal/pkg/ai-proxy/route/route.go:93]: why `+=2`

File: pkg/reverseproxy/reverseproxy.go [hunk 16]
```diff
@@ -319,6 +319,10 @@ func (p *ReverseProxy) serveHTTP(rw http.ResponseWriter, req *http.Request) {
 			}
 		}
 	}
+	// reset content-length
+	contentLength := infor.BodyBuffer().Len()
+	outreq.ContentLength = int64(contentLength)
+	outreq.Header.Set("Content-Length", strconv.Itoa(contentLength))
```
Comment (sfwn) [inline pkg/reverseproxy/reverseproxy.go:325]: use const in pkg or apistructs.

**Review by sfwn — CHANGES_REQUESTED** [review]
Remove `main` file.

File: internal/apps/ai-proxy/filters/azure-director/filter.go (hunk 6 above)
Comment (dspo) [reply internal/apps/ai-proxy/filters/azure-director/filter.go:271]: 这里最初是用一个 `map[string]func(...)` 维护了一个方法列表, 从配置文件读取到要调用的方法, 从 map 中取出这个方法来调用。后来每增删加一个方法， 都要去修改这个 map，所以重构成了用反射获取同一签名的系列方法。编写代码时打印了方法签名，后来不需要打印了，就删掉了。不用改成 debug level 了。

File: internal/apps/ai-proxy/providers/metrics/collector.go (hunk 12 above)
Comment (dspo) [reply internal/apps/ai-proxy/providers/metrics/collector.go:86]: removed it

File: internal/apps/ai-proxy/vars/consts_test.go (hunk 14 above)
Comment (dspo) [reply internal/apps/ai-proxy/vars/consts_test.go:26]: removed

File: pkg/reverseproxy/reverseproxy.go (hunk 16 above)
Comment (dspo) [reply pkg/reverseproxy/reverseproxy.go:325]: resolved

File: internal/apps/ai-proxy/handlers/handler_sessions.go (hunk 8 above)
Comment (dspo) [reply internal/apps/ai-proxy/handlers/handler_sessions.go:207]: resolve it on next pr

File: internal/apps/ai-proxy/handlers/handler_sessions.go (hunk 9 above)
Comment (dspo) [reply internal/apps/ai-proxy/handlers/handler_sessions.go:230]: on next pr

File: internal/apps/ai-proxy/provider.go (hunk 11 above)
Comment (dspo) [reply internal/apps/ai-proxy/provider.go:203]: removed

**Review by sfwn — APPROVED** [review]

**Pull request merged**